### PR TITLE
planner: avoid using index_merge when there are multiple table filters

### LINKED ISSUE #22105: The result is not as expected when using the hint of use_index_merge
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

[1.txt](https://github.com/pingcap/tidb/files/5754106/1.txt)
download the file
```
use test;
source "path of the file";
SELECT /*+ use_index_merge(t1)*/ COUNT(*) FROM t1 WHERE (key4=42 AND key6 IS NOT NULL) OR (key1=4 AND key3=6); 
```

### 2. What did you expect to see? (Required)
31
### 3. What did you see instead (Required)
0
### 4. What is your TiDB version? (Required)
release-5.0-rc

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve?

Issue Number: close #22105  

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:
Avoid using index merge when there are multiple table filters.

How it Works:

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects

N/A

### Release note 

- Avoid using index merge when there are multiple table filters.

### CODE REVIEW

File: planner/core/stats.go [hunk 1]
```diff
@@ -554,8 +557,15 @@ func (ds *DataSource) buildIndexMergeOrPath(partialPaths []*util.AccessPath, cur
 	indexMergePath := &util.AccessPath{PartialIndexPaths: partialPaths}
 	indexMergePath.TableFilters = append(indexMergePath.TableFilters, ds.pushedDownConds[:current]...)
 	indexMergePath.TableFilters = append(indexMergePath.TableFilters, ds.pushedDownConds[current+1:]...)
+	tableFilterCnt := 0
 	for _, path := range partialPaths {
+		// IndexMerge should not be used when the SQL is like 'select x from t WHERE (key1=1 AND key2=2) OR (key1=4 AND key3=6);'.
+		// Check issue https://github.com/pingcap/tidb/issues/22105 for details.
```
Comment (eurekaka) [inline planner/core/stats.go:563]: Actually we can use IndexMerge for this case, as long as we keep the whole filter `(key1=1 and key2=2) or (key1=4 and key3=6)` in the selection.

**Review by eurekaka — COMMENTED** [review]
LGTM. We can fix the bug quickly by this PR and optimize the plan later.

**Review by qw4990 — APPROVED** [review]
LGTM

**Pull request merged**